Precise per-component ETH-denominated rewards tracking

zah opened the pull request:
This is an alternative take on https://github.com/status-im/nimbus-eth2/pull/3107
that aims for more minimal interventions in the spec modules at the expense of
duplicating more of the spec logic in ncli_db.

zah committed:
Precise per-component ETH-denominated rewards tracking

This is an alternative take on https://github.com/status-im/nimbus-eth2/pull/3107
that aims for more minimal interventions in the spec modules at the expense of
duplicating more of the spec logic in ncli_db.

arnetheduck commented on beacon_chain/spec/state_transition_block.nim:
```diff
@@ -425,6 +425,21 @@ proc process_operations(cfg: RuntimeConfig,
+      get_base_reward_per_increment(total_active_balance) * total_active_increments
+    max_participant_rewards =
+      total_base_rewards * SYNC_REWARD_WEIGHT div WEIGHT_DENOMINATOR div SLOTS_PER_EPOCH
+  return max_participant_rewards div SYNC_COMMITTEE_SIZE
```
```suggestion
  max_participant_rewards div SYNC_COMMITTEE_SIZE
```

arnetheduck commented on beacon_chain/beacon_chain_db.nim:
```diff
@@ -76,7 +76,7 @@ type
     ## between different versions of the client and accidentally using an old
     ## database.
-    db: SqStoreRef
+    db*: SqStoreRef
```
expose a `readOnly` template instead of making the db public?

arnetheduck commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -119,6 +119,39 @@ func initiate_validator_exit*(cfg: RuntimeConfig, state: var ForkyBeaconState,
+  elif state is merge.BeaconState:
+      validator_effective_balance div MIN_SLASHING_PENALTY_QUOTIENT_MERGE
+  else:
+    raiseAssert "invalid BeaconState type"
```
```suggestion
    static: raiseAssert "invalid BeaconState type"
```

... or overloads, then each could get its own link in the doc comment

arnetheduck commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -119,6 +119,39 @@ func initiate_validator_exit*(cfg: RuntimeConfig, state: var ForkyBeaconState,
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/phase0/beacon-chain.md#slash_validator
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/altair/beacon-chain.md#modified-slash_validator
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/merge/beacon-chain.md#modified-slash_validator
+proc get_whistleblower_reward*(validator_effective_balance: Gwei): Gwei =
```
```suggestion
func get_whistleblower_reward*(validator_effective_balance: Gwei): Gwei =
```

arnetheduck commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -119,6 +119,39 @@ func initiate_validator_exit*(cfg: RuntimeConfig, state: var ForkyBeaconState,
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/phase0/beacon-chain.md#slash_validator
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/altair/beacon-chain.md#modified-slash_validator
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/merge/beacon-chain.md#modified-slash_validator
+proc get_slashing_penalty*(state: ForkyBeaconState,
```
```suggestion
func get_slashing_penalty*(state: ForkyBeaconState,
```

arnetheduck commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -119,6 +119,39 @@ func initiate_validator_exit*(cfg: RuntimeConfig, state: var ForkyBeaconState,
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/phase0/beacon-chain.md#slash_validator
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/altair/beacon-chain.md#modified-slash_validator
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/merge/beacon-chain.md#modified-slash_validator
+proc get_proposer_reward(state: ForkyBeaconState, whistleblower_reward: Gwei): Gwei =
```
```suggestion
func get_proposer_reward(state: ForkyBeaconState, whistleblower_reward: Gwei): Gwei =
```

arnetheduck commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -119,6 +119,39 @@ func initiate_validator_exit*(cfg: RuntimeConfig, state: var ForkyBeaconState,
+  elif state is altair.BeaconState or state is merge.BeaconState:
+    whistleblower_reward * PROPOSER_WEIGHT div WEIGHT_DENOMINATOR
+  else:
+    raiseAssert "invalid BeaconState type"
```
```suggestion
    static: raiseAssert "invalid BeaconState type"
```

arnetheduck commented on beacon_chain/spec/state_transition_epoch.nim:
```diff
@@ -827,34 +851,55 @@ func process_registry_updates*(
+    elif state is merge.BeaconState:
+      PROPORTIONAL_SLASHING_MULTIPLIER_MERGE
+    else:
+      raiseAssert "process_slashings: incorrect BeaconState type")
```
```suggestion
      static: raiseAssert "process_slashings: incorrect BeaconState type")
```

arnetheduck commented on beacon_chain/spec/state_transition_epoch.nim:
```diff
@@ -827,34 +851,55 @@ func process_registry_updates*(
+      PROPORTIONAL_SLASHING_MULTIPLIER_MERGE
+    else:
+      raiseAssert "process_slashings: incorrect BeaconState type")
+  return min(sum(state.slashings.data) * multiplier, total_balance)
```
```suggestion
  min(sum(state.slashings.data) * multiplier, total_balance)
```

arnetheduck commented on beacon_chain/spec/state_transition_epoch.nim:
```diff
@@ -827,34 +851,55 @@ func process_registry_updates*(
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/altair/beacon-chain.md#slashings
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/specs/merge/beacon-chain.md#slashings
+func slashing_penalty_applies*(validator: Validator, epoch: Epoch): bool =
+  return validator.slashed and
```
```suggestion
  validator.slashed and
```

arnetheduck commented on beacon_chain/spec/state_transition_epoch.nim:
```diff
@@ -827,34 +851,55 @@ func process_registry_updates*(
+                                                # numerator to avoid uint64 overflow
+  let penalty_numerator = validator.effective_balance div increment *
+                          adjusted_total_slashing_balance
+  return penalty_numerator div total_balance * increment
```
```suggestion
  penalty_numerator div total_balance * increment
```

arnetheduck reviewed (approved):
nice, this is a lot more clean - ie the spec code is better than it was and the validator code remains pretty straightforward over-all - it also becomes cheap / easy to modify the validatorDb command further without touching the more sensitive spec code which is great :+1:

arnetheduck commented:
hm, actually, in the future, it's not impossible that we can move some of the code to a separate profit computation module - I can see this being useful in implementing additional REST requests or in the GUI

zah commented:
This also took a lot more time to develop and debug. It will likely continue to do so in the future after any new hard-forks tweaking the rewards. I would personally value any savings in the aggregated development time higher than the aspiration to keep the spec code "clean". It seems trivial to me to ignore the `when` blocks introduced in the spec code during audits, but if you have strong feelings about this, I can live with this more complicated version.

I also looked briefly at the possibility to compress the output by writing only "source" information which the reward info can be derived in Jupyter, but proved to be quite tricky. I've found something like a sweet spot between writing speed, storage requirements and report production speed by employing compressed CSV files
.

arnetheduck commented:
Indeed, I do believe it's important for the long-term quality of the spec code to not make these concessions, based on its relative importance - I appreciate the extra effort spent on keeping things clean.

zah committed:
More efficient reward data persistance; Address review comments

The new format is based on compressed CSV files in two channels:

* Detailed per-epoch data
* Aggregated "daily" summaries

The use of append-only CSV file speeds up significantly the epoch
processing speed during data generation. The use of compression
results in smaller storage requirements overall. The use of the
aggregated files has a very minor cost in both CPU and storage,
but leads to near interactive speed for report generation.

Other changes:

- Implemented support for graceful shut downs to avoid corrupting
  the saved files.

- Fixed a memory leak caused by lacking `StateCache` clean up on each
  iteration.

- Addressed review comments

- Moved the rewards and penalties calculation code in a separate module

Required invasive changes to existing modules:

- The `data` field of the `KeyedBlockRef` type is made public to be used
  by the validator rewards monitor's Chain DAG update procedure.

- The `getForkedBlock` procedure from the `blockchain_dag.nim` module
  is made public to be used by the validator rewards monitor's Chain DAG
  update procedure.
